src/unix/darwin.c

```diff
@@ -180,11 +185,144 @@ int uv_uptime(double* uptime) {
+                        RTLD_LAZY | RTLD_LOCAL);
+
+  if (core_foundation_handle == NULL || iokit_handle == NULL) {
+    goto out;
+  }
+
+#define V(handle, symbol)                                                     \
+  do {                                                                        \
+    *(void **)(&p ## symbol) = dlsym((handle), #symbol);                      \
+    if (p ## symbol == NULL)                                                  \
+      goto out;                                                               \
+  }                                                                           \
+  while (0)
+  V(iokit_handle, IOMasterPort);
+  V(iokit_handle, IOServiceMatching);
+  V(iokit_handle, IOServiceGetMatchingServices);
+  V(iokit_handle, IOIteratorNext);
+  V(iokit_handle, IOObjectRelease);
+  V(iokit_handle, IORegistryEntryCreateCFProperty);
+  V(core_foundation_handle, CFStringCreateWithCString);
+  V(core_foundation_handle, CFStringGetSystemEncoding);
+  V(core_foundation_handle, CFDataGetBytePtr);
+  V(core_foundation_handle, CFDataGetLength);
+  V(core_foundation_handle, CFDataGetBytes);
+  V(core_foundation_handle, CFRelease);
+#undef V
+
+#define S(s) pCFStringCreateWithCString(NULL, (s), kCFStringEncodingUTF8)
+  io_iterator_t it;
+  io_object_t service;
+
+  kr = pIOMasterPort(MACH_PORT_NULL, &mach_port);
+  // TODO(evanlucas) handle non KERN_SUCCESS return values
+  CFMutableDictionaryRef classes_to_match
+      = pIOServiceMatching("IOPlatformDevice");
+  kr = pIOServiceGetMatchingServices(mach_port, classes_to_match, &it);
+  // TODO(evanlucas) handle non KERN_SUCCESS return values
+  service = pIOIteratorNext(it);
+
+  CFStringRef device_type_str = S("device_type");
+  CFStringRef clock_frequency_str = S("clock-frequency");
+
+  do {
+    if (service) {
+      CFDataRef data = NULL;
+      data = pIORegistryEntryCreateCFProperty(service,
+                                              device_type_str,
+                                              NULL,
+                                              0);
+      if (data) {
+        const UInt8 *raw = pCFDataGetBytePtr(data);
+        if (strncmp((char *)raw, "cpu", 3) == 0
+          || strncmp((char *)raw, "processor", 9) == 0) {
+
+          data = pIORegistryEntryCreateCFProperty(service,
+                                                  clock_frequency_str,
+                                                  NULL,
+                                                  0);
+          if (data) {
+            uint32_t freq;
+            CFIndex len = pCFDataGetLength(data);
+            CFRange range;
+            range.location = 0;
+            range.length = len;
+
+            pCFDataGetBytes(data, range, (UInt8 *)&freq);
+            *speed = freq;
+            pCFRelease(data);
+            break;
+          }
+        }
+        pCFRelease(data);
+      }
+    }
+    service = pIOIteratorNext(it);
+  } while (service != 0);
```

Comment:
The logic is now:
```c
do {
  if (service != 0) {
    // ...
  }
  service = pIOIteratorNext(it);
} while (service != 0);
```
That can just be a simple while loop, right?
```c
while (service != 0)  {
  // ...
  service = pIOIteratorNext(it);
}
```

Reply:
Yes this was much easier. Good call